storage/swift_storage.py

```diff
@@ -1,253 +1,254 @@
-                                           username=username, password=password,
-                                           api_key=api_key, tenant_id=tenant_id)
-            context.auth_endpoint = auth_endpoint
-            context.authenticate()
-            return context
-
-        context = timeout(DEFAULT_SWIFT_TIMEOUT, create_swift_context_and_authenticate)
-        self._cloudfiles = context.get_client("swift", region, public=public)
-        self._cloudfiles.timeout = DEFAULT_SWIFT_TIMEOUT
-
-    def _get_container_and_object_names(self):
-        _, container_name = self._parsed_storage_uri.netloc.split("@")
-        object_name = self._parsed_storage_uri.path[1:]
-        return container_name, object_name
 
-    def _list_container_objects(self, container_name, prefix):
-        container_objects = self._cloudfiles.list_container_objects(container_name, prefix=prefix)
-        directories = []
-        files = []
+    auth_endpoint: Optional[str] = None
 
-        for container_object in container_objects:
-            if container_object.content_type == "application/directory":
-                directories.append(container_object)
-            else:
-                files.append(container_object)
+    # cache get connections
+    def get_connection(self) -> swiftclient.client.Connection:
+        if not hasattr(self, "_connection"):
+            query = dict(parse_qsl(self._parsed_storage_uri.query))
 
-        return directories, files
+            auth_endpoint = query.get("auth_endpoint")
+            if auth_endpoint is None:
+                raise SwiftStorageError(f"Required field is missing: auth_endpoint")
 
-    def save_to_filename(self, file_path):
-        with open(file_path, "wb") as output_fp:
-            self.save_to_file(output_fp)
+            self.tenant_id = query.get("tenant_id")
+            if self.tenant_id is None:
+                raise SwiftStorageError(f"Required field is missing: tenant_id")
 
-    def save_to_file(self, out_file):
-        self._authenticate()
-        container_name, object_name = self._get_container_and_object_names()
+            region_name = query.get("region")
+            if region_name is None:
+                raise SwiftStorageError(f"Required field is missing: region_name")
 
-        for chunk in self._cloudfiles.fetch_object(
-                container_name, object_name, chunk_size=_LARGE_CHUNK):
-            out_file.write(chunk)
+            self.download_url_key = query.get("download_url_key")
 
-    def save_to_directory(self, destination_directory):
-        self._authenticate()
-        container_name, prefix = self._get_container_and_object_names()
+            os_options = {
+                "tenant_id": self.tenant_id,
+                "region_name": region_name
+            }
 
-        directories, files = self._list_container_objects(container_name, prefix)
+            auth, _ = self._parsed_storage_uri.netloc.split("@")
```

Comment:
I'm not sure why it was done this way originally, but should we be using the username/password fields returned by `urlparse`, instead of manually parsing the `netloc` ourselves?